File: frontend/src/types/taxon.ts
export const Rank = {
    KINGDOM: "Kingdom",
    DIVISION: "Division",
    CLASS: "Class",
    ORDER: "Order",
    FAMILY: "Family",
    GENUS: "Genus",
    SPECIES: "Species",
} as const;

export type Rank = typeof Rank[keyof typeof Rank];

export interface Taxon {
    id: string;
    name: string;
    rank: Rank;
    description?: string;
    parent_id?: string;
    parent?: Taxon;
    children?: Taxon[];
    created_at?: string;
    updated_at?: string;
}

export type TaxonResponse = Taxon;

export interface TaxonTree extends Taxon {
    children: TaxonTree[];
}

export interface TaxonCreate {
    name: string;
    rank: Rank;
    description?: string;
    parent_id?: string;
}

export interface TaxonUpdate {
    name?: string;
    rank?: Rank;
    description?: string;
    parent_id?: string;
}

File: backend/app/schemas/taxon.py
from typing import List, Optional, ForwardRef
import uuid
from pydantic import BaseModel, ConfigDict

from app.models.enums import Rank

# Forward reference for recursive model
TaxonRecursive = ForwardRef('TaxonRecursive')

class TaxonBase(BaseModel):
    name: str
    rank: Rank
    description: Optional[str] = None
    parent_id: Optional[uuid.UUID] = None

class TaxonCreate(TaxonBase):
    pass

class TaxonUpdate(TaxonBase):
    name: Optional[str] = None
    rank: Optional[Rank] = None

class TaxonResponse(TaxonBase):
    id: uuid.UUID
    model_config = ConfigDict(from_attributes=True)

class TaxonTree(TaxonResponse):
    children: List["TaxonTree"] = []
    
    model_config = ConfigDict(from_attributes=True)

# Resolve forward reference
TaxonTree.model_rebuild()
